README.md
# Logo Learning

Version 0.1.0

Motivating Learning Programming Visually and Intuitively

The Logo Learning tool is a visual turtle Logo interpreter built behind a website user interface. It is designed to provide an intuitive introduction to some basic programming concepts essentially converting the natural way people think into essential programming ideas. The tool takes three different forms of input including abstract Logo language code, intermediate button inputs, and direct drawing with the mouse. These different forms of input make it easy to control the program as a beginner, yet all of them eventually boil down to Logo code which allows people to understand how to use programmatic input and build more complex constructs. The main motivation for learning comes from the specifically designed challenges that present paths for the users to draw with the Logo turtle that teach programming concepts like loops, functions, and parameters. The whole tool has tutorials that automatically pop up to inform the user on screen how to use everything and that also function as interactive documentation and user guides, taking away more of the complexity of learning.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
### Changed
### Deleted

## [0.1.0] - 2020-04-16
### Added
- The CHANGELOG!

### Changed
- Canvas size to 300
- Positioning of divs to fit the page


canvas.js
function CanvasState(htmlCanvas,turtle,previewTurtle){
	this.canvas = htmlCanvas;
	this.ctxt = htmlCanvas.getContext('2d');
	this.turtle = turtle;
	this.previewTurtle = previewTurtle; 
	this.points = [];

	this.challengesObj = new challenges();

	this.draw();
}

CanvasState.prototype.drawTurtle = function(turtleObj){
	var x = turtleObj.x();
	var y = turtleObj.y();
	var angle = turtleObj.angle();
	this.ctxt.fillStyle=turtleObj.isPreview?"#999999":HACKER?"#FFFFFF":"#000000";
	
	this.ctxt.beginPath();
	xdisp = Math.cos(angle)*5;
	ydisp = Math.sin(angle)*5;
	this.ctxt.moveTo(x+xdisp, y+ydisp);
	xdisp = Math.cos(angle+Math.PI*7/6)*5;
	ydisp = Math.sin(angle+Math.PI*7/6)*5;
	this.ctxt.lineTo(x+xdisp,y+ydisp);
	xdisp = Math.cos(angle+Math.PI*5/6)*5;
	ydisp = Math.sin(angle+Math.PI*5/6)*5;
	this.ctxt.lineTo(x+xdisp,y+ydisp);
	this.ctxt.fill();
}

CanvasState.prototype.draw = function(){
	this.ctxt.clearRect(0,0,this.canvas.width,this.canvas.height);

	//TODO: may want to move the checkSovled fxn call
	this.challengesObj.checkSolved(this);
	this.challengesObj.draw(this);

	this.ctxt.strokeStyle = "#999999";
	for(var i=0;i<this.previewTurtle.lines.length;i++){
		this.ctxt.beginPath();
		this.ctxt.moveTo(this.previewTurtle.lines[i][0][0],this.previewTurtle.lines[i][0][1]);
		this.ctxt.lineTo(this.previewTurtle.lines[i][1][0],this.previewTurtle.lines[i][1][1]);
		this.ctxt.stroke();
	}

	this.ctxt.strokeStyle = HACKER?"#FFFFFF":"#000000";
	for(var i=0;i<this.turtle.lines.length;i++){
		this.ctxt.beginPath();
		this.ctxt.moveTo(this.turtle.lines[i][0][0],this.turtle.lines[i][0][1]);
		this.ctxt.lineTo(this.turtle.lines[i][1][0],this.turtle.lines[i][1][1]);
		this.ctxt.stroke();
	}

	this.drawTurtle(this.previewTurtle);
	this.drawTurtle(this.turtle);
}

CanvasState.prototype.addLine = function(startPos,endPos){
	this.turtle.lines.push([startPos,endPos]);

	var vec = xy2vec(startPos[0],startPos[1],endPos[0],endPos[1]);
	sx = Math.cos(vec.angle);
	sy = Math.sin(vec.angle);

	for(var d=0,x=startPos[0],y=startPos[1];d<vec.mag;d++){
		this.points.push([x,y]);
		x+=sx;
		y+=sy;
	}
}

CanvasState.prototype.addPreviewLine = function(startPos,endPos){
	this.previewTurtle.lines.push([startPos,endPos]);
}

CanvasState.prototype.clear = function(){
	this.points = [];
	this.turtle.lines = [];
	this.ctxt.clearRect(0,0,this.canvas.width,this.canvas.height);
	this.draw();
}

CanvasState.prototype.clearPreview = function(){
	this.previewTurtle.lines = [];
	//this.draw();
}

CanvasState.prototype.changeChallenge = function(i){
	this.challengesObj.setChallenge(i);
	this.draw();
}

function setUpCanvas(){
	var canvas = document.createElement('canvas');

	var dim = 300;
	canvas.setAttribute('width',dim)
	canvas.setAttribute('height',dim)
	canvas.setAttribute('id','canvas');
    
    var cDiv = document.getElementById('canvasDiv');
    cDiv.insertBefore(canvas,cDiv.firstChild);
    cDiv.insertBefore(document.createElement('br'),cDiv.children[1]);

    canvas.addEventListener('click', line2click);

    var t = new Turtle(dim/2,dim/2,false);
    var pt = new Turtle(dim/2,dim/2,true);

    var state = new CanvasState(canvas,t,pt);

    return state;
}

function getCursorPosition(e) {
    var x;
    var y;

    var rect = canvas.getBoundingClientRect();
    x = Math.round((e.clientX-rect.left)/rect.width*canvas.width);
	y = Math.round((e.clientY-rect.top)/rect.height*canvas.height);
	x += e.pageX-e.clientX+rect.left;
	y += e.pageY-e.clientY+rect.top;
    
    return [x, y];
}

function line2click(e){
  context = this.getContext('2d');

  var p = getCursorPosition(e);
  mx = p[0] - this.offsetLeft;
  my = p[1] - this.offsetTop;

  var vec=xy2vec(canvasState.turtle.x(),canvasState.turtle.y(),mx,my);

  degreeTurn = Math.round( (vec.angle - canvasState.turtle.angle())/Math.PI*180 )%360;
  disp = Math.round(vec.mag);

  turnCmd = degreeTurn<0?'lt '+(-degreeTurn):'rt '+degreeTurn;

  //console.log('rt '+degreeTurn+' fd '+disp);
  runSource(turnCmd+' fd '+ disp);
}

function xy2vec(x1,y1,x2,y2){
	var dx,dy;
	dx = x2-x1;
	dy = y2-y1;
	var result = {};
	result.angle = Math.atan2(dy,dx);
	result.mag = Math.sqrt(dx*dx+dy*dy);
	return result;
}

CanvasState.prototype.resetPreview = function(){
  this.clearPreview();
  this.previewTurtle.x(this.turtle.x());
  this.previewTurtle.y(this.turtle.y());
  this.previewTurtle.angle(this.turtle.angle());
  canvasState.draw();
};


turtle.js
function Turtle(x,y,isPreview){
	this._x=x;
	this._y=y;
	this._angle=0;
	this._penDown=true;
	this.lines=[];
	this.isPreview=isPreview;
	this.logoFunctions={};

	this.logoFunctions.fd = function(n){
		if(arguments.length<1) throw new MissingArgErr('fd',1);
		this.forwardBy(n);
	};

	this.logoFunctions.bk = function(n){
		if(arguments.length<1) throw new MissingArgErr('bk',1);
		this.forwardBy(-n);
	};

	this.logoFunctions.rt = function(n){
		if(arguments.length<1) throw new MissingArgErr('rt',1);
		this.turnBy(n);
	};

	this.logoFunctions.lt = function(n){
		if(arguments.length<1) throw new MissingArgErr('lt',1);
		this.turnBy(-n)
	};

	this.logoFunctions.pd = function(){
		this.penDown(true);
	};

	this.logoFunctions.pu = function(){
		this.penDown(false);
	};

	this.logoFunctions.clearscreen = function(){
		this.clear();
	};

	this.logoFunctions.if = function(cond,block){
		if(arguments.length<2) throw new MissingArgErr('if',2);
		if(cond){block();}else{undefined;}
	};

	this.logoFunctions.repeat = function(n,block){
		if(arguments.length<2) throw new MissingArgErr('repeat',2);
		for(var i=0;i<n;i++){
			block();
		}
	};
};

function UndefFuncErr(functionName){ 
	this.message = 'Don\'t know how to "'+functionName+'"';
	this.stack = (new Error()).stack; 
};

UndefFuncErr.prototype = new Error;
UndefFuncErr.prototype.name='UndefinedFunctionError';

function MissingArgErr(functionName,n){
	this.message = 'Can\'t "'+functionName+'" needs '+n+' argument(s)';
	this.stack = (new Error()).stack;
};

MissingArgErr.prototype = new Error;
MissingArgErr.prototype.name = 'MissingArgumentError';

Turtle.prototype.lookup = function(name){
	if(this.logoFunctions[name]){
		return this.logoFunctions[name];
	}else{
		throw new UndefFuncErr(name);
	}
};

Turtle.prototype.x = function(val){
	if(arguments.length) this._x=val;
	else return this._x;
};

Turtle.prototype.y = function(val){
	if(arguments.length) this._y=val;
	else return this._y;
};

Turtle.prototype.angle = function(val){
	if(arguments.length) this._angle=val;
	else return this._angle;
};

Turtle.prototype.penDown = function(val){
	if(arguments.length) this._penDown=val;
	else return this._penDown;
};

Turtle.prototype.printPositionAndAngle = function(){
  text = Math.round(this.x()) + ',';
  text += Math.round(this.y()) + ',';
  text += ((Math.round(this.angle()*180/Math.PI)%360)+360)%360;
  document.getElementById('locationText').innerHTML = text;
};

// ******************START TURTLE COMMANDS**************************
Turtle.prototype.forwardBy = function(n){        
  var x1,y1,x2,y2;
  var start = [];
  var end = [];
  start[0] = this.x();
  start[1] = this.y();
  end[0] = start[0] + n*Math.cos(this.angle());
  end[1] = start[1] + n*Math.sin(this.angle());
  this.x(end[0]);
  this.y(end[1]);

  if(this.penDown()){
    if(!this.isPreview){
      canvasState.addLine(start,end);
    }else{
      canvasState.addPreviewLine(start,end);
    }
  }
  canvasState.draw();

  if(!this.isPreview) this.printPositionAndAngle();
}            

Turtle.prototype.turnBy = function(n){
  this._angle += (n * Math.PI / 180);
  canvasState.draw();
  if(!this.isPreview) this.printPositionAndAngle();
} 

Turtle.prototype.setPen = function(writeBool){
  this.penDown(writeBool);
}

Turtle.prototype.clear = function(){
  this.x(150);
  this.y(150);
  this.angle(0);
  if(!this.isPreview){
    canvasState.clear();
  }else{
    canvasState.clearPreview();
  }
}


grammar.js
Turtle rt: n [self turnBy: n].
Turtle lt: n [self turnBy: n negated].
Turtle fd: n [self forwardBy: n].
Turtle bk: n [self forwardBy: n negated].
Turtle pu    [self setPenDown: false].
Turtle pd    [self setPenDown: true].

ometa LogoTranslator : Parser {
  name         ::= <spaces> <firstAndRest #letter #letterOrDigit>:xs => [xs squish join: ''].
  cmdName      ::= <name>:n
                     ?[n ~= 'to'] ?[n ~= 'end'] ?[n ~= 'output']     => [n].
  number       ::= <spaces> <digit>+.
  arg          ::= <token ':'> <name>.
  cmds         ::= <cmd>*:xs                                         => [xs delimWith: ';'].
  block        ::= <token '['> <cmds>:xs <token ']'>                 => ['(function() {',, xs,, '})'].
  primExpr     ::= <arg> | <number> | <block>
                 | <token '('> (<expr> | <cmd>):x <token ')'>        => [x].
  mulExpr      ::= <mulExpr>:x <token '*'> <primExpr>:y              => [x,, '*',, y]
                 | <mulExpr>:x <token '/'> <primExpr>:y              => [x,, '/',, y]
                 | <primExpr>.
  addExpr      ::= <addExpr>:x <token '+'> <mulExpr>:y               => [x,, '+',, y]
                 | <addExpr>:x <token '-'> <mulExpr>:y               => [x,, '-',, y]
                 | <mulExpr>.
  relExpr      ::= <addExpr>:x <token '<'>  <addExpr>:y              => [x,, '<',, y]
                 | <addExpr>:x <token '<='> <addExpr>:y              => [x,, '<=',, y]
                 | <addExpr>:x <token '>'>  <addExpr>:y              => [x,, '>',, y]
                 | <addExpr>:x <token '>='> <addExpr>:y              => [x,, '>=',, y]
                 | <addExpr>.
  expr         ::= <relExpr>.
  cmd          ::= <token 'output'> <expr>:x                         => ['return ',, x]
                 | <cmdName>:n <expr>*:args                          => ['$elf.performwithArguments("',, n,, '", [',,
                                                                             (args delimWith: ','),,
                                                                          '])'].
  decl         ::= <token 'to'> <cmdName>:n <arg>*:args
                     <cmds>:body  <token 'end'>                      => ['$elf.',, n,, ' = ',,
                                                                           'function(',, (args delimWith: ','),, ') {',,
                                                                             body,,
                                                                           '}'].
  topLevelCmd  ::= <decl> | <cmd>.
  topLevelCmds ::= <topLevelCmd>*:xs <spaces> <end>                  => [xs := '(function() { var $elf = this; ',,
                                                                                 (xs delimWith: ';'),, '})'.
                                                                         xs squish join: ''].
}.

Turtle eval: code [
  | trans |
  trans := LogoTranslator matchAll: code with: #topLevelCmds.
  self at: #LOGODOIT put: (eval value: trans).
  self LOGODOIT
].

Turtle if: cond : aBlock [
  cond ifTrue: aBlock
].

smiley := Turtle new.
smiley eval: '

  to spiral :size :angle
    if :size < 100 [
      fd :size
      rt :angle
      spiral :size + 2 :angle
    ]
  end
  spiral 0 89

'

      

tutorial.js
var DIMENSION = 30
var DIMENSION_STR= String(DIMENSION)
var lastElement;

function startIntro(){
  var intro = introJs();
  intro.setOption('showBullets',false);
  intro.oncomplete(function(){
  	runSource('clearscreen');
  });
  intro.onchange(function(targetElement){
  	if(this._direction=='forward'){
  		switch(this._currentStep){
  			case 4:
  			case 8:
  				lastElement.click();
  				break;
  			case 7:
  				lastElement.firstElementChild.value='bk 100';
  				break;
  			case 10:
  				runSource('rt 111\nfd 16');
  				break;
  		}
  	}
  	lastElement = targetElement;
  });
  intro.setOptions({
	steps: [
	  { 
		intro: 'This is the Logo Learning tool. Click next to learn about it or skip if you\'ve aleady seen the tutorial.'
	  },
	  {
		element: '#canvas',
		intro: 'The turtle on the screen is your main tool. It can move forwards, backwards, and rotate leaving a line wherever it goes.',
		position: 'top'
	  },
	  {
		element: '#toolbarDiv',
		intro: 'There are three methods of input. First, is the button toolbar.',
		position: 'left'
	  },
	  {
		element: '#fdBtn',
		intro: 'Click a button to move the turtle.',
		position: 'left'
	  },
	  {
		element: '#canvas',
		intro: 'Notice the turtle moved forward in the direction it was pointed.',
		position: 'top'
	  },
	  {
		element: '#source',
		intro: 'Notice the button click also generated the command for moving the turtle forward.',
		position: 'right'
	  },
	  {
		element: '#sourceCodeDiv',
		intro: 'The second method input is the soure code area. Here commands can be edited.',
		position: 'right'
	  },	  
	  {
		element: '#doItBtn',
		intro: 'To run some code from here click the "do it" button.',
		position: 'bottom'
	  },
	  {
		element: '#canvas',
		intro: 'Notice the turtle moved 100 units backwards',
		position: 'top'
	  },
	  {
		element: '#canvas',
		intro: 'The third method of input is the canvas itself. Clicking causes the turtle to turn and go to that location.',
		position: 'top'
	  },
	  {
		element: '#source',
		intro: 'Observe that two commands were generated this time in the source code area.',
		position: 'right'
	  },
	  {
	  	element: '#challengeDiv',
	  	intro: 'Once you can use all three inputs try to complete the challenges by selecting a category.',
		position: 'top'
	  },
	  {
	  	element: '#basicsGroupDiv',
	  	intro: 'Start with basics and work your way through them all.',
		position: 'top'
	  },
	  {
	  	element: '#csBtn',
	  	intro: 'Clicking "clear screen" will reset the turtle. Draw some more or try one of the challenges.',
		position: 'left'
	  }
	]
  });

  intro.start();
}

function startChallengeIntro(name){
  var intro = introJs();
  intro.setOption('showBullets',false);
  switch(name){
  	case 'basics':
	  intro.onchange(function(targetElement){
	  	if(this._direction=='forward'){
	  		switch(this._currentStep){
	  			case 1:
	  				runSource('clearscreen');
	  				break;
	  			case 2:
	  			case 4:
	  			case 7:
	  				lastElement.click();
	  				break;
	  			case 6:
	  				console.log(lastElement);
	  				lastElement.value='fd 30';
	  				break;

	  		}
	  	}
	  	lastElement = targetElement;
	  });
	  intro.setOptions({
		steps: [
		  { 
			intro: 'The challenges present a path to complete. Move the turtle through the entire path to complete the challenge.'
		  },
		  {
			element: '#challenge0Btn',
			intro: 'Select a challenge by clicking on it to display it in the canvas.',
			position: 'bottom'
		  },
		  {
			element: '#canvas',
			intro: 'The blue path is the one you must complete.',
			position: 'bottom'
		  },
		  {
			element: '#fdBtn',
			intro: 'It appears the turtle needs to move forward.',
			position: 'left'
		  },
		  {
			element: '#canvas',
			intro: 'It appears the turtle needs to go a bit farther.',
			position: 'top'
		  },
		  {
			element: '#source',
			intro: 'Edit the generated code to move the remaining distance',
			position: 'right'
		  },
		  {
			element: '#doItBtn',
			intro: 'Press "do it" to run the newly typed code.',
			position: 'right'
		  },
		  {
			element: '#canvas',
			intro: 'Notice the path changed colors to indicate it was completed.',
			position: 'top'
		  },
		  {
			element: '#basicsGroupDiv',
			intro: 'Try to complete the rest of the challenges here.',
			position: 'top'
		  }
		]
	  });
	  break;
	case 'repeat':
	  intro.onchange(function(targetElement){
	  	if(this._direction=='forward'){
	  		switch(this._currentStep){
	  			case 0:
	  				runSource('clearscreen');
	  				document.getElementById('challenge3Btn').click();
	  				break;
	  			case 1:
	  				setSource('fd ' + DIMENSION_STR + '\nlt 120');
	  				break;
	  			case 4:
	  				lastElement.click();
	  				break;
	  			case 5:
	  				lastElement.click();
	  		}
	  	}
	  	lastElement = targetElement;
	  });
	  intro.setOptions({
		steps: [
		  { 
			intro: 'This set of challenges makes use of the repeat commands.'
		  },
		  {
			element: '#canvas',
			intro: 'The first challenge requires you to draw a line and turn. This can be acheived doing the same thing twice.',
			position: 'top'
		  },
		  {
			element: '#source',
			intro: 'These two commands draw the line and turn the proper amount.',
			position: 'right'
		  },
		  {
			element: '#repeatBtn',
			intro: 'Click the repeat button then another command to do it multiple times.',
			position: 'left'
		  },
		  {
			element: '#doItBtn',
			intro: 'Next click the "do it" button. In the dialog enter 2 then press OK to run the commands twice.',
			position: 'right'
		  },
		  {
			element: '#source',
			intro: 'Notice the command that was generated. Anything in the [] is done the specified number of times.',
			position: 'right'
		  },
		  {
			element: '#repeatGroupDiv',
			intro: 'Try to complete the rest of the challenges using repeat.',
			position: 'top'
		  }

		]
	  });
	  break;
	case 'functions':
      intro.onchange(function(targetElement){
	  	if(this._direction=='forward'){
	  		switch(this._currentStep){
	  			case 1:
	  				runSource('clearscreen');
	  				document.getElementById('challenge11Btn').click();
	  				break;
	  			case 3:
	  			case 12:
	  				if(!RECORDING) lastElement.click();
	  				break;
	  			case 4:
	  			case 6:
	  			case 9:
	  			case 11:
	  			case 13:
	  			case 14:
	  			case 19:
	  				lastElement.click();
	  				break;
	  			case 7:
	  				document.getElementById('source').value = 'to function'+MYFUNNUM+'\nfd ' + DIMENSION_STR + '\nlt 120\nend\n';
	  				break;
	  			case 8:
	  				if(RECORDING) lastElement.click();
	  				break;
	  			case 15:
	  				document.getElementById('source').value = 'to triangle\nrepeat 3 [ function'+(MYFUNNUM-1)+' ]\nend\n';
	  				break;
	  			case 17:
	  				if(RECORDING) lastElement.click();
	  				setSource('triangle\nfd ' + DIMENSION_STR + '\ntriangle');
	  				break;
	  		}
	  	}
	  	lastElement = targetElement;
	  });
	  intro.setOptions({
		steps: [
		  { 
			intro: 'These challenges use functions with the "to ... end" syntax.'
		  },
		  {
			element: '#canvas',
			intro: 'The first challenge requires you to draw 4 triangles. You can do this by recording funtions.',
			position: 'top'
		  },
		  {
			element: '#recordBtn',
			intro: 'Press "record" to create a function. A function will be a set of commands you can run with one command.',
			position: 'left'
		  },
		  {
			element: '#fdBtn',
			intro: 'Once you press "record", it adds each command you enter to the function until you press "stop"',
			position:'left'
		  },
		  {
			element: '#source',
			intro: 'Notice the to function# ... end was added to the source area.',
			position: 'right'
		  },		  
		  {//5
			element: '#ltBtn',
			intro: 'To finish one side of a triangle you need to turn as well as go forward.',
			position: 'left'
		  },		  
		  {
			element: '#source',
			intro: 'Edit the code to make the turtle move and turn the right amount for a triangle.',
			position: 'right'
		  },		  
		  {
			element: '#recordBtn',
			intro: 'Press "stop" to exit recording mode and to store the function.',
			position: 'left'
		  },
		  {
		  	element: '#playBtn',
			intro: 'Once you have recorded a function you can run it by pressing "play"',
			position: 'left'
		  },
		  {
			element: '#canvas',
			intro: 'Notice the one command moves the turtle forward and turns it.',
			position: 'top'
		  },
		  {//10
			element: '#csBtn',
			intro: 'Clear the screen to erase the scratch work.',
			position: 'left'
		  },
		  {
			element: '#recordBtn',
			intro: 'Next we will try to record a full triangle function.',
			position: 'left'
		  },
		  {
			element: '#repeatBtn',
			intro: 'It will include repeating the last function for a side 3 times.',
			position: 'left'
		  },
		  {
			element: '#playBtn',
			intro: 'Click "play" and then enter 3 to repeat the side funtion 3 times.',
			position: 'left'
		  },
		  {
			element: '#source',
			intro: 'You can change the name of the function to help you remember what it does.',
			position: 'right'
		  },
		  {//15
			element: '#source',
			intro: 'Notice the function is now called "triangle". To call it you just type the word "triangle" in the source area.',
			position: 'right'
		  },
		  {
			element: '#recordBtn',
			intro: 'Pressing "stop" will exit recording mode and store the function.',
			position: 'left'
		  },
		  {
			element: '#source',
			intro: 'Now the single command "triangle" will draw the entire triangle for you like the other commands.',
			position: 'right'
		  },
		  {
		  	element: '#doItBtn',
		  	intro: 'Press "do it" to draw the triangle in the canvas',
		  	position: 'right'
		  },
		  {
			element: '#canvas',
			intro: 'Notice the single command now draws a whole triangle.',
			position: 'top'
		  },
		  {//20
			element: '#functionsGroupDiv',
			intro: 'Try to complete the challenges using functions.',
			position: 'top'
		  }
		]
	  });
	  break;
	case 'parameters':
      intro.onchange(function(targetElement){
	  	if(this._direction=='forward'){
	  		switch(this._currentStep){
	  			case 1:
	  				runSource('clearscreen');
	  				document.getElementById('challenge14Btn').click();
	  				break;
	  			case 2:
	  				var s = document.getElementById('source');
	  				s.value = 'to triangle\n  repeat 3 [\n    fd ' + DIMENSION_STR + '\n    lt 120  ]\nend\n';
	  				break;
	  			case 3:
	  				var s = document.getElementById('source');
	  				s.value = 'to triangle :size\n  repeat 3 [\n    fd ' + DIMENSION_STR + '\n    lt 120  ]\nend\n';
	  				setCaretSelection(s,12,17);
	  				break;
	  			case 4:
	  				var s = document.getElementById('source');
	  				s.value = 'to triangle :size\n  repeat 3 [\n    fd :size\n    lt 120  ]\nend\n';
	  				setCaretSelection(s,38,43);
	  				break;
	  			case 6:
	  				lastElement.click();
	  				var s = document.getElementById('source');
	  				s.value = 'triangle ' + DIMENSION_STR + '\n';
	  				break;
	  			case 8:
	  				lastElement.click();
	  			case 9:
	  				var s = document.getElementById('source');
	  				s.value = 'triangle ' + String(DIMENSION+10) + '\ntriangle ' + String(DIMENSION+20) + '\n';
	  		}
	  	}
	  	lastElement = targetElement;
	  });
	  intro.setOptions({
		steps: [
		  { 
			intro: 'These challenges use function parameters/arguments.'
		  },
		  {
			element: '#canvas',
			intro: 'The first challenge requires drawing three different sized triangles.',
			position: 'top'
		  },
		  {
			element: '#source',
			intro: 'Here is a function to create a fixed size ' + DIMENSION_STR + ' triangle. This triangle length cannot be changed right now.',
			position: 'right'
		  },
		  {
			element: '#source',
			intro: 'To add a parameter type ":" followed by a name after the function name. Parameters enable changing something about a function each time it is used.',
			position: 'right'
		  },
		  {
			element: '#source',
			intro: 'To use the parameter in the function put ":<parameterName>" wherever you want to use it. In this case the ":size" parameter determines how far forward the turtle moves.',
			position: 'right'
		  }
		  ,
		  {
			element: '#doItBtn',
			intro: 'Press "do it" to store the new triangle function with a parameter.',
			position: 'right'
		  },
		  {
			element: '#source',
			intro: 'Try it out by typing "triangle" then the parameter value. A parameter is required now for triangle to work.',
			position: 'right'
		  },
		  {
			element: '#doItBtn',
			intro: 'Press "do it" to draw your triangle with size ' + DIMENSION_STR + '.',
			position: 'right'
		  },
		  {
			element: '#canvas',
			intro: 'Notice the function drew the triangle with the right size.',
			position: 'top'
		  }
		  ,
		  {
			element: '#source',
			intro: 'Try drawing the triangles with different sizes.',
			position: 'right'
		  },
		  {
			element: '#parametersGroupDiv',
			intro: 'Try to complete all the challenges using parameters.',
			position: 'top'
		  }
		]
	  });
	  break;
  }

  intro.start();
}


challenges.js
var HACKER = false;
var MIDDLE = 150;
var THICKNESS = 5;

function foundHacker(){
	HACKER=true;
	var fileref=document.createElement("link")
        fileref.setAttribute("rel", "stylesheet")
        fileref.setAttribute("type", "text/css")
        fileref.setAttribute("href", "hack.css")
        document.getElementsByTagName("head")[0].appendChild(fileref)
}

function challenges(){
	this.challengeArray = [];
	this.challengeIdx = 0;

	this.challengeGroupsSolved = 0;

	this.loadPaths();
	this.addButtons();
}

function toggleDiv(name){
	var e = document.getElementById(name+'Div');
	if(e.style.display == 'block'){
	  	e.style.display = 'none';
	}else{
		startChallengeIntro(name);
	  	e.style.display = 'block';
	}
}

challenges.prototype.addButtonGroup = function(name,startIdx,endIdx,cDiv){
	var i;
	var div;
	var gDiv;
	var toggle;

	gDiv = document.createElement('div');
	gDiv.id = name+'GroupDiv';
	gDiv.style.border = 'solid #00BFFF';
	gDiv.style['margin-bottom'] = '3pt';

	toggle = document.createElement('button');
	toggle.setAttribute('class','text-button');
	toggle.id = name + 'Toggle';
	toggle.innerHTML = name;
	toggle.onclick = function(){toggleDiv(name);};
	gDiv.appendChild(toggle);

	div = document.createElement('div');
	div.setAttribute('id',name+'Div');
	div.style.display = 'none'
	gDiv.appendChild(div);
	cDiv.appendChild(gDiv);
	
	for(i=startIdx;i<endIdx+1;i++){
		this.createButton(i,div);
	}
}

challenges.prototype.addButtons = function(){
	var cDiv = document.getElementById('challengeDiv');

	this.addButtonGroup('basics',0,2,cDiv);

	this.addButtonGroup('repeat',3,10,cDiv);

	this.addButtonGroup('functions',11,13,cDiv);

	this.addButtonGroup('parameters',14,16,cDiv);
}

challenges.prototype.createButton = function(i,parent){
	var btn = document.createElement('button');
	var t = document.createTextNode('Challenge '+i);
	btn.id = 'challenge'+i+'Btn';
	btn.appendChild(t);
	btn.setAttribute('onclick', 'canvasState.changeChallenge('+i+')');
	btn.setAttribute('class', 'challenge-button');
	parent.appendChild(btn);
}

challenges.prototype.checkGroupSolved = function(i){
	var floor,top,name;
	if(i<3){
		floor=0;
		top=2;
		name='basics';
	}else if(i<11){
		floor=3;
		top=10;
		name='repeat';	
	}else if(i<14){
		floor=11;
		top=13;
		name='functions';
	}else{
		floor=14;
		top=16;
		name='parameters';
	}
	for(var j=floor;j<top+1;j++){
		if(!this.challengeArray[j].solved) break;
	}
	var c='chartreuse';
	if(j===top+1){
		c='#FFFF66';
		if(++this.challengeGroupsSolved===4){
			foundHacker();
		}
	}
	document.getElementById(name+'GroupDiv').style['borderColor'] = c;

}

challenges.prototype.loadPaths = function(){
	var paths = [];
	var dim = 30;

	//***repeat challenges***
	paths.push({start:[MIDDLE,MIDDLE],end:[MIDDLE+dim,MIDDLE]});
	this.addChallenge(paths);

	//paths = [];
	paths=this.regularPolygon(4,dim,MIDDLE,MIDDLE);
	this.addChallenge(paths.slice(0,2))
	this.addChallenge(paths);

	paths=this.regularPolygon(3,dim,MIDDLE,MIDDLE,true);
	this.addChallenge(paths.slice(0,2))
	this.addChallenge(paths);

	var sides = [6,12,24];
	for(var sIdx=0;sIdx<sides.length;sIdx++){
		paths=this.regularPolygon(sides[sIdx],sIdx==2?dim/2:dim,MIDDLE,MIDDLE,sIdx==2);
		this.addChallenge(paths.slice(0,2))
		this.addChallenge(paths);
	}

	//***function to..end challenges***
	//adding multiple triangles
	var i;
	paths=[];
	for(i=0;i<4;i++){
		paths=paths.concat(this.regularPolygon(3,dim,MIDDLE+dim*i,MIDDLE,true));
	}
	this.addChallenge(paths);

	//adding multiple squares
	var i;
	paths=[];
	for(i=0;i<4;i++){
		paths=paths.concat(this.regularPolygon(4,dim,MIDDLE+dim*i,MIDDLE));
	}
	this.addChallenge(paths);

	paths=[];
	for(i=0;i<4;i++){
		paths=paths.concat(this.regularPolygon(3,dim,MIDDLE+dim*i,MIDDLE,true));
		paths=paths.concat(this.regularPolygon(4,dim,MIDDLE+dim*i,MIDDLE));
	}
	this.addChallenge(paths);

	//***parameter to :arg...end challenges***
	paths=[];
	var size;
	var offset=0;
	for(i=0;i<3;i++){
		size=10*i+dim;
		paths=paths.concat(this.regularPolygon(3,size,MIDDLE+offset,MIDDLE,true));
		offset+=size;
	}
	this.addChallenge(paths);

	paths=[];
	var size;
	var offset=0;
	for(i=0;i<3;i++){
		size=10*i+dim;
		paths=paths.concat(this.regularPolygon(4,size,MIDDLE+offset,MIDDLE));
		offset+=size;
	}
	this.addChallenge(paths);

	paths=[];
	var size;
	var offset=0;
	for(i=0;i<3;i++){
		size=10*i+dim;
		paths=paths.concat(this.regularPolygon(3,size,MIDDLE+offset,MIDDLE,true));
		paths=paths.concat(this.regularPolygon(4,size,MIDDLE+offset,MIDDLE));
		offset+=size;
	}
	this.addChallenge(paths);
}

challenges.prototype.regularPolygon = function(sides,len,x,y,optionalDirection){
	var paths=[];
	var dx,dy;
	var angle=2*Math.PI/sides;
	if(optionalDirection) angle=-angle;
	for(var i=0;i<sides;i++){
		dx=Math.cos(i*angle)*len;
		dy=Math.sin(i*angle)*len;
		paths.push({start:[x,y],end:[x+dx,y+dy]});
		x+=dx;
		y+=dy;
	}
	return paths;
}

challenges.prototype.addChallenge = function(paths){
	theChallenge = {};
	theChallenge.paths = paths;
	theChallenge.solved = false;
	theChallenge.points = [];
	for(var i=0;i<theChallenge.paths.length;i++){
		var start = theChallenge.paths[i].start;
		var end = theChallenge.paths[i].end;
		var angle,mag;
		var dx,dy;

		var vec = xy2vec(start[0],start[1],end[0],end[1]);

		angle = vec.angle;
		mag=vec.mag;
		dx=Math.cos(angle);
		dy=Math.sin(angle);

		for(var x=start[0], y=start[1], m=0;m<mag;m++){
			theChallenge.points.push([x,y]);
			x+=dx;
			y+=dy;
		}
	}
	this.challengeArray.push(theChallenge);
}

challenges.prototype.getChallenge = function(){
	return this.challengeArray[this.challengeIdx];
}

challenges.prototype.setChallenge = function(i){
	this.challengeIdx = i;
}

challenges.prototype.draw = function(cs){
	var theChallenge = this.getChallenge();

	var start;
	var angle;
	var mag;
	var dx,dy;
	
	cs.ctxt.fillStyle=theChallenge.solved?'#FFFF66':'#00BFFF';
	for(var i=0;i<theChallenge.points.length;i++){
		cs.ctxt.beginPath();
		cs.ctxt.arc(theChallenge.points[i][0],theChallenge.points[i][1],THICKNESS,0,2*Math.PI);
		cs.ctxt.fill();
	}
	cs.ctxt.fillStyle='#000000';
}

challenges.prototype.checkSolved = function(cs){
	var theChallenge = this.getChallenge();

	if(theChallenge.solved) return;

	for(var i=0;i<theChallenge.points.length;i++){
		var p1 = theChallenge.points[i];
		for(var j=0;j<cs.points.length;j++){
			if(distBetween(p1,cs.points[j])<=THICKNESS*2){
				break;
			}
		}
		if(j==cs.points.length) return;
	}
	theChallenge.solved = true;
	document.getElementById('challenge'+this.challengeIdx+'Btn').setAttribute('class','challenge-button-solved');
	this.checkGroupSolved(this.challengeIdx);
	return;
}

function distBetween(p1,p2){
	dx=p1[0]-p2[0];
	dy=p1[1]-p2[1];
	return Math.sqrt(dx*dx+dy*dy);
}
